pkg/controller/uniteddeployment/uniteddeployment_controller.go

```diff
@@ -257,3 +259,115 @@ func (r *ReconcileUnitedDeployment) classifySubsetBySubsetName(ud *appsv1alpha1.
 	}
 	return mapping
 }
+
+func (r *ReconcileUnitedDeployment) updateStatus(instance *appsv1alpha1.UnitedDeployment, oldStatus *appsv1alpha1.UnitedDeploymentStatus, nameToSubset *map[string]*Subset, nextReplicas, nextPartition *map[string]int32, currentRevision, updatedRevision *appsv1.ControllerRevision, collisionCount int32) (reconcile.Result, error) {
+	newStatus := r.calculateStatus(instance, nameToSubset, nextReplicas, nextPartition, currentRevision, updatedRevision, collisionCount)
+	_, err := r.updateUnitedDeployment(instance, oldStatus, newStatus)
+	return reconcile.Result{}, err
+}
+
+func (r *ReconcileUnitedDeployment) calculateStatus(ud *appsv1alpha1.UnitedDeployment, nameToSubset *map[string]*Subset, nextReplicas, nextPartition *map[string]int32, currentRevision, updatedRevision *appsv1.ControllerRevision, collisionCount int32) *appsv1alpha1.UnitedDeploymentStatus {
+	newStatus := ud.Status
+
+	expectedRevision := currentRevision.Name
+	if updatedRevision != nil {
+		expectedRevision = updatedRevision.Name
+	}
+
+	var replicas int32
+	var readyReplicas int32
+	var updatedReplicas int32
+	var updatedReadyReplicas int32
+
+	// sync from status
+	for _, subset := range *nameToSubset {
+		subsetReplicas, subsetReadyReplicas, subsetUpdatedReplicas, subsetUpdatedReadyReplicas := replicasStatusFn(subset, expectedRevision)
+		replicas += subsetReplicas
+		readyReplicas += subsetReadyReplicas
+		updatedReplicas += subsetUpdatedReplicas
+		updatedReadyReplicas += subsetUpdatedReadyReplicas
+	}
+
+	newStatus.Replicas = replicas
+	newStatus.ReadyReplicas = readyReplicas
+	newStatus.UpdatedReplicas = updatedReplicas
+	newStatus.UpdatedReadyReplicas = updatedReadyReplicas
+
+	newStatus.SubsetReplicas = *nextReplicas
+
+	if newStatus.CurrentRevision == "" {
+		// init with current revision
+		newStatus.CurrentRevision = currentRevision.Name
+	}
+
+	if newStatus.UpdateStatus == nil {
+		newStatus.UpdateStatus = &appsv1alpha1.UpdateStatus{}
+	}
+
+	newStatus.UpdateStatus.UpdatedRevision = expectedRevision
+	newStatus.UpdateStatus.CurrentPartitions = *nextPartition
+
+	if newStatus.UpdateStatus != nil && newStatus.UpdateStatus.UpdatedRevision != newStatus.CurrentRevision && newStatus.UpdatedReadyReplicas >= newStatus.Replicas {
```

Comment:
This check newStatus.UpdateStatus != nil  seems to be unnecessary.

Reply:
done